components/proxy/src/main/kotlin/com/hotels/styx/routing/db/StyxObjectStore.kt

```diff
@@ -16,27 +16,136 @@
+import org.reactivestreams.Publisher
+import reactor.core.publisher.Flux
 import java.util.Optional
-import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.CopyOnWriteArrayList
+import java.util.concurrent.Executors
+import java.util.concurrent.atomic.AtomicReference
 
 /**
  * Styx Route Database.
  */
 class StyxObjectStore<T> : ObjectStore<T> {
-    private val objects = ConcurrentHashMap<String, Record<T>>();
+    private val objects: AtomicReference<PMap<String, Record<T>>> = AtomicReference(HashTreePMap.empty())
+    private val watchers = CopyOnWriteArrayList<ChangeWatcher<T>>()
 
-    override fun get(name: String?): Optional<T> {
-        return Optional.ofNullable(objects.get(name))
+    companion object {
+        private val executor = Executors.newSingleThreadExecutor()
+    }
+
+    /**
+     * Retrieves an object from this object store.
+     *
+     * Returns `Optional.empty` when an object with `key` doesn't exist.
+     *
+     * This method is thread safe. It can be called simultaneously from many threads.
+     *
+     * @property key object name
+     * @return an optional object
+     */
+
+    override fun get(name: String): Optional<T> {
+        return Optional.ofNullable(objects().get(name))
                 .map { it.payload }
     }
 
+    /**
+     * Inserts a new object in object store.
+     *
+     * If an object already exist with same `key`, then the previous object is
+     * replaced with new one.
+     *
+     * Notifies watchers.
+     *
+     * This method is thread safe. It can be called simultaneously from many threads.
+     *
+     * @property key object name
+     * @property payload the object itself
+     */
+    fun insert(key: String, payload: T) {
+        insert(key, setOf(), payload)
+    }
+
+    /**
+     * Removes an object from this object store.
+     *
+     * Watchers are notified after successful removal.
+     *
+     * If `key` doesn't exist, then nothing happens and watchers are not notified.
+     *
+     * This method is thread safe. It can be called simultaneously from many threads.
+     *
+     * @property key object name
+     * @property payload the object itself
+     */
+    fun remove(key: T) {
+        queue {
+            val nextVersion = objects().minus(key)
+
+            if (nextVersion != objects()) {
+                objects.set(nextVersion)
+                notifyWatchers(nextVersion)
+            }
+        }
+    }
+
+    /**
+     * Returns a Publisher that emits an event at any modification.
+     *
+     * Watch activates on subscription only.
+     * Watch removed on unsubscription.
+     */
+    fun watch(): Publisher<ObjectStore<T>> {
+        return Flux.push { sink ->
+            val watcher: ChangeWatcher<T> = { sink.next(it) }
+
+            sink.onDispose {
+                watchers.remove(watcher)
+            }
+
+            watchers.add(watcher)
+
+            queue {
+                sink.next(ObjectStore { key ->
```

Comment:
This watch() method is doing quite a few things so I would probably extract functions that can have a descriptive name, especially for this last bit.

Reply:
Hmm... I could do something like this:

    fun watch(): Publisher<ObjectStore<T>> {
        return Flux.push { sink ->
            addWatcher(sink)

            queue {
                emitInitialSnapshot(sink)
            }
        }
    }

But then again `addWatcher` would not only create and add a watcher, but it would also configure `sink.onDispose` callback. Seems bit contrived. 

But I could still factor out `emitInitialSnapshot` method as above.